### ISSUE TITLE
Feature Request: Allow instances to be reused, but also garbage collected

### ISSUE THREAD

**Issue by ankushg** [InsertKoinIO/koin#1378]
It's great that we have some flexibility between `factory` and `singleton`!

At the same time, we have many classes that could be reused if they already exist, without a strong need to only have one single instance. 

- With `singleton`, we find ourselves wasting memory by retaining such classes far beyond when we need them
- With `factory`, we find ourselves wasting memory by creating too many simultaneous instances of classes that _could have_ be shared

It would be great if we could tell Koin to 
- reuse an existing instance if it _already has_ one, but also 
- give Koin the freedom to let that instance be garbage collected whenever it's no longer needed and recreated in the future

This is similar to Dagger's `@Reusable` scope.

**Comment by arnaudgiuliani** [comment]
Yep, something like with weak references to garbage it if nobody has nay reference on it 🤔

**Comment by lcszc** [comment]
Any plans on when this will get added? It's been 1y+

[Swinject](https://github.com/Swinject/Swinject/blob/master/Documentation/ObjectScopes.md) already has this feature as well and I believe it is a must-have, otherwise we would need to be hacking around scopes.

**Comment by arnaudgiuliani** [comment]
yes, it's something that need a bit of rework for internal engine

**Comment by tamimattafi** [comment]
Are there any updates on this?

**Comment by arnaudgiuliani** [comment]
Let's put that in a KFIP process. Could Thread local / weak reference with dedicated factory. If we can design it for 4.1 that's cool.

**Comment by arnaudgiuliani** [comment]
Due to work on Koin Compiler Plugin , need to delay for 4.3 👍

**Comment by Nailik** [comment]
@arnaudgiuliani can you please update the Labels and remove wontfix
